cli/popper/utils.py

```diff
@@ -175,32 +278,27 @@ def b(t):
 
     if log_file:
         if verbose:
-            if debug:
-                info('\nDEBUG: Creating file for combined stdout/stderr\n')
+            debug_log('\nCreating file for combined stdout/stderr')
```

Comment:
We can also use this opportunity to have `logging` deal with log files. Instead of us having to create the log files, we can instruct `logging` to create them. This will simplify things greatly, as we only need to do `log.info()`.

If we do the above, then we need to check the value of `log_file` and setup `logging` accordingly so that it also creates log files.

Reply:
This means that the logic described at the beginning of the function (where we describe the four cases), can be handled entirely by making use of `logging`, and we only create a single stream of output that we pass to `log.info()`.

Reply:
Can you explain a bit about how we are generating logs? Specifically where in the code are we writing to the files? Thanks

Reply:
sure. We open the file(s) for writing in [this block](https://github.com/systemslab/popper/blob/master/cli/popper/utils.py#L176-L185), and write to them [here](https://github.com/systemslab/popper/blob/master/cli/popper/utils.py#L213).

However, now that we are using `logging`, we can simplify things:

  * expose a `--log` flag to `popper run` that allows users to specify where to write a single log file to. We can have it disabled by default (empty string or `None`).
  * when `--log` is not empty, we pass it to the `setup_logging` function so that we configure the `logging` custom handler in a way that we create a log file in the path specified by the value of the flag
  * then, the runners (host, docker and singularity) only need to write output using the `logging` API and do not need to deal with files

ihth

Reply:
So I wanted to clarify, in this Logic mentioned
```
    # the main logic is the following:
    #
    # 1) verbose=False and log_file=None
    #      ==> don't write anything to stdout/log
    # 2) verbose=True and log_file=None
    #      ==> combine stdout/stderr in the same stream and print it to stdout
    # 3) verbose=False and log_file not None
    #      ==> write two files, one .out and one .err
    # 4) verbose=True and log_file not None
    #      ==> combine stdout/stderr, write to stdout and to a SINGLE log file
```
*  Regarding point 1 and 2, are these conditions ever meet? Because the only call to `exec_cmd` [here](https://github.com/systemslab/popper/blob/d719f9276dbbc7c4eb56af4a78b05d12c7b68668/cli/popper/gha.py#L710-L713) is taking `log_file=self.log_filename` and `self.log_filename` is always defined in the `ActionRunner`, so the condition `log_file=None` is never met.
* For point 3 I was thinking I will use the current level being used ( default: `ACTION_INFO` , `--debug`: `DEBUG` and `--quiet`: `INFO`) for `.out` file and the level `ERROR` for `.err` file.
* For point 4, the level `DEBUG` for `.log` files.

Reply:
> @ivotron just wanted to confirm, is [this](https://github.com/systemslab/popper/blob/d719f9276dbbc7c4eb56af4a78b05d12c7b68668/cli/popper/gha.py#L530) `pu.info(b(l))` where we print our action info?

yeah, that's for the dockerrunner. In the case of the singularity runner, it is [here](https://github.com/systemslab/popper/blob/d719f9276dbbc7c4eb56af4a78b05d12c7b68668/cli/popper/gha.py#L672).

For the docker runner, I already took out the code that was dealing with writing to log files. This can be simplified further by removing the conditional for when `quiet` and `debug` are given.

For singularity, we can simplify the logic of the code significantly, as we can remove conditionals that setup log files, as well as those that deal with `quiet`, `verbose` and `debug`.

Reply:
couple of things that can help simplify the code:

  * as this function is now only used by the `HostRunner`, we can move it from `utils.py` to `gha.py` 
  * drop the `verbose` print to `log.debug`, `log.info` and `log.actioninfo` (assuming that's the name of the new level)
  * drop `log_file` as we will now deal with files via the custom `logging` handler (i.e. there's no need to create files here in this function, as that'll be done by the handler)

ihth

Reply:
So are we still generating two separate files `.err` and `.out`/ or one `.log` file? And are we mirroring the output thrown to the stdout in the logs or will it be different?

Reply:
in general, the use of `logging` means two things:
 1. the entire popper codebase will **only** use `log.debug`, `log.actioninfo`, `log.info`, etc.. to log messages. This means that we don't need to add special logic (the conditionals that check for the values of `debug` or `quiet`) to determine when these get logged, nor the code base needs to directly create files. The main goal of using a logging library is to delegate those tasks to the it.
 2. we can configure the behavior of the `logging` handler so that it knows which messages are logged (the logging level) and to where they get sent to (stdout/stderr, files, a database, etc.). This configuration can be done based on the flags that users passed to `popper run`:

      * if `--log` is given, then instruct the `logging` custom handler to create a **single** file.
      * if `--quiet` is given, then increase the level so that `ACTION_INFO` messages are turned off (as `ACTION_INFO` would be the default level).
      * if `--debug` is given, then decrease the level so that `DEBUG` messages are also logged.

Reply:
So I added the `--log-file` flag  which takes and generates a logfile, and it has the same level as the `stdout` , so that whatever is shown on the console is written to a logfile. 
If the flag is not present I create a logfile in the path `popper_logs\` and a file with the name `log` + `timestamp` in that directory.